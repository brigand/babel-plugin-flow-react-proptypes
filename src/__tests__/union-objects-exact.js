const babel = require('babel-core');
const content = `
var React = require('react');


type U = {
  a: string
};
type V = {|
  b: number
|};

type FooProps = {
  someprop: U | V,
};

export type ExactFooProps = {|
  someprop: U | V,
|};

class Foo extends React.Component {
  props: FooProps;
}

class ExactFoo extends React.Component {
  props: ExactFooProps;
}


`;

it('union-objects-exact', () => {
  const res = babel.transform(content, {
    babelrc: false,
    presets: ['es2015', 'stage-1', 'react'],
    plugins: ['syntax-flow', require('../')],
  }).code;
  expect(res).toMatchSnapshot();
});
