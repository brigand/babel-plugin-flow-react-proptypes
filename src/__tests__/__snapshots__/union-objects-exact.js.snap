// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`union-objects-exact 1`] = `
"'use strict';

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError(\\"Cannot call a class as a function\\"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError(\\"this hasn't been initialised - super() hasn't been called\\"); } return call && (typeof call === \\"object\\" || typeof call === \\"function\\") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== \\"function\\" && superClass !== null) { throw new TypeError(\\"Super expression must either be null or a function, not \\" + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var React = require('react');

var babelPluginFlowReactPropTypes_proptype_ExactFooProps = {
  someprop: require('prop-types').oneOfType([require('prop-types').shape({
    a: require('prop-types').string.isRequired
  }), require('prop-types').shape({
    b: require('prop-types').number.isRequired,
    __exact__: function __exact__(values, prop, displayName) {
      var props = {
        b: true
      };
      var extra = [];

      for (var k in values) {
        if (values.hasOwnProperty(k) && !props.hasOwnProperty(k)) {
          extra.push(k);
        }
      }

      if (extra.length > 0) {
        return new Error('Invalid additional prop(s) ' + JSON.stringify(extra));
      }
    }
  })]).isRequired
};
if (typeof exports !== 'undefined') Object.defineProperty(exports, 'babelPluginFlowReactPropTypes_proptype_ExactFooProps', {
  value: babelPluginFlowReactPropTypes_proptype_ExactFooProps,
  configurable: true,
  enumerable: true
});

var Foo = function (_React$Component) {
  _inherits(Foo, _React$Component);

  function Foo() {
    _classCallCheck(this, Foo);

    return _possibleConstructorReturn(this, (Foo.__proto__ || Object.getPrototypeOf(Foo)).apply(this, arguments));
  }

  return Foo;
}(React.Component);

Foo.propTypes = {
  someprop: require('prop-types').oneOfType([require('prop-types').shape({
    a: require('prop-types').string.isRequired
  }), require('prop-types').shape({
    b: require('prop-types').number.isRequired,
    __exact__: function __exact__(values, prop, displayName) {
      var props = {
        b: true
      };
      var extra = [];

      for (var k in values) {
        if (values.hasOwnProperty(k) && !props.hasOwnProperty(k)) {
          extra.push(k);
        }
      }

      if (extra.length > 0) {
        return new Error('Invalid additional prop(s) ' + JSON.stringify(extra));
      }
    }
  })]).isRequired
};

var ExactFoo = function (_React$Component2) {
  _inherits(ExactFoo, _React$Component2);

  function ExactFoo() {
    _classCallCheck(this, ExactFoo);

    return _possibleConstructorReturn(this, (ExactFoo.__proto__ || Object.getPrototypeOf(ExactFoo)).apply(this, arguments));
  }

  return ExactFoo;
}(React.Component);

ExactFoo.propTypes = {
  someprop: require('prop-types').oneOfType([require('prop-types').shape({
    a: require('prop-types').string.isRequired
  }), require('prop-types').shape({
    b: require('prop-types').number.isRequired,
    __exact__: function __exact__(values, prop, displayName) {
      var props = {
        b: true
      };
      var extra = [];

      for (var k in values) {
        if (values.hasOwnProperty(k) && !props.hasOwnProperty(k)) {
          extra.push(k);
        }
      }

      if (extra.length > 0) {
        return new Error('Invalid additional prop(s) ' + JSON.stringify(extra));
      }
    }
  })]).isRequired
};"
`;
